### PR TITLE
Support vpe

### PR DESCRIPTION
**Pull request by haim-kermany**
added support for vpe
basically, copying the behavior of ni-vsi
Adi, please review the vpcmodel / ibmvpc code
Ziv, please review the rest

### CODE REVIEW

File: pkg/drawio/createMapFile.go [hunk 1]
```diff
@@ -12,17 +12,15 @@ import (
...
-	ShowNIIcon bool
-	Nodes      []TreeNodeInterface
+	FipXOffset           int
+	FipYOffset           int
+	MiniIconXOffset      int
+	MiniIconYOffset      int
+	MiniIconSize         int
+	RootID               uint
+	IDsPrefix            string
+	canTypeHaveAMiniIcon map[reflect.Type]bool
```
Comment (zivnevo) [inline pkg/drawio/createMapFile.go:22]: Why is this here?
Doesn't it fit better in `styles.go` where all the other maps with `reflect.Type` key are?

File: pkg/drawio/treeNodeInterface.go [hunk 2]
```diff
@@ -27,7 +27,7 @@ package drawio
 // the structs abstractSquareTreeNode, abstractIconTreeNode, abstractLineTreeNode contains abstractTreeNode
 // All structs representing a Square (VpcTreeNode, ZoneTreeNode, SubnetTreeNode...) contains abstractIconTreeNode
 // All structs representing an icons (NITreeNode, GatewayTreeNode, UserTreeNode...) contains abstractIconTreeNode
-// All structs representing a line (VsiLineTreeNode, ConnectivityTreeNode) contains abstractIconTreeNode
+// All structs representing a line (LogicalLineTreeNode, ConnectivityTreeNode) contains abstractIconTreeNode
```
Comment (zivnevo) [inline pkg/drawio/treeNodeInterface.go:30]:
```suggestion
// All structs representing a line (LogicalLineTreeNode, ConnectivityTreeNode) contain abstractLineTreeNode
```
Please review the comments above as well

File: pkg/ibmvpc/ibmDrawioGenerator.go [hunk 3]
```diff
@@ -80,6 +69,19 @@ func (v *Vsi) GenerateDrawioTreeNode(gen *vpcmodel.DrawioGenerator) drawio.TreeN
...
 }
 
+func (v *Vpe) GenerateDrawioTreeNode(gen *vpcmodel.DrawioGenerator) drawio.TreeNodeInterface {
+	if len(v.Nodes()) == 0 {
+		return nil
+	}
+	resIPs := []drawio.TreeNodeInterface{}
+	for _, ni := range v.Nodes() {
+		resIPs = append(resIPs, gen.TreeNode(ni))
+	}
```
Comment (zivnevo) [inline pkg/ibmvpc/ibmDrawioGenerator.go:79]:
```suggestion
	for _, resIP := range v.Nodes() {
		resIPs = append(resIPs, gen.TreeNode(resIP))
	}
```

File: pkg/ibmvpc/ibmDrawioGenerator.go [hunk 4]
```diff
@@ -80,6 +69,19 @@ func (v *Vsi) GenerateDrawioTreeNode(gen *vpcmodel.DrawioGenerator) drawio.TreeN
 	return nil
 }
 
+func (v *Vpe) GenerateDrawioTreeNode(gen *vpcmodel.DrawioGenerator) drawio.TreeNodeInterface {
+	if len(v.Nodes()) == 0 {
+		return nil
+	}
+	resIPs := []drawio.TreeNodeInterface{}
```
Comment (zivnevo) [inline pkg/ibmvpc/ibmDrawioGenerator.go:76]: Pre-allocate memory with `make`

File: pkg/drawio/iconTreeNode.go [hunk 5]
```diff
@@ -4,6 +4,7 @@ type IconTreeNodeInterface interface {
 	TreeNodeInterface
 	RouterID() uint
 	SG() SquareTreeNodeInterface
+	setSG(SquareTreeNodeInterface)
```
Comment (zivnevo) [inline pkg/drawio/iconTreeNode.go:7]: Maybe add a TODO here for supporting multiple SGs per icon

File: pkg/drawio/layout.go [hunk 6]
```diff
@@ -138,33 +138,36 @@ func sortGroupSquareBySize(groups []SquareTreeNodeInterface) []SquareTreeNodeInt
 //	         - else its visibility is connectedPoint
 func (ly *layoutS) calcGroupsVisibility(subnet SquareTreeNodeInterface) {
```
Comment (zivnevo) [inline pkg/drawio/layout.go:139]:
```suggestion
func calcGroupsVisibility(subnet SquareTreeNodeInterface) {
```
Function can be static

File: pkg/drawio/layout.go [hunk 7]
```diff
@@ -138,33 +138,36 @@ func sortGroupSquareBySize(groups []SquareTreeNodeInterface) []SquareTreeNodeInt
...
 	sortedBySizeGroups := sortGroupSquareBySize(subnet.(*SubnetTreeNode).groupSquares)
-	iconSquareGroups := map[IconTreeNodeInterface]map[SquareTreeNodeInterface]bool{}
+	iconShownSquareGroups := map[IconTreeNodeInterface]map[SquareTreeNodeInterface]bool{}
 	for _, groupS := range sortedBySizeGroups {
 		group := groupS.(*GroupSquareTreeNode)
-		if len(group.groupedIcons) == len(subnet.(*SubnetTreeNode).NIs()) {
-			group.setVisibility(theSubnet)
-			continue
-		}
 		groupedIconsFormerGroups := map[SquareTreeNodeInterface]bool{}
```
Comment (zivnevo) [inline pkg/drawio/layout.go:144]: `groupedIconsFormerGroups` doesn't really have to be a map. Consider converting it to a slice.

**Review by zivnevo — CHANGES_REQUESTED** [review]

**Review by adisos — COMMENTED** [review]
Looks good.

File: pkg/drawio/createMapFile.go (hunk 1 above)
Comment (haim-kermany) [reply pkg/drawio/createMapFile.go]: I rearranged the code, having the styles in a separate struct

File: pkg/drawio/treeNodeInterface.go (hunk 2 above)
Comment (haim-kermany) [reply pkg/drawio/treeNodeInterface.go]: done

File: pkg/ibmvpc/ibmDrawioGenerator.go (hunk 3 above)
Comment (haim-kermany) [reply pkg/ibmvpc/ibmDrawioGenerator.go]: done

File: pkg/ibmvpc/ibmDrawioGenerator.go (hunk 4 above)
Comment (haim-kermany) [reply pkg/ibmvpc/ibmDrawioGenerator.go]: done

File: pkg/drawio/iconTreeNode.go (hunk 5 above)
Comment (haim-kermany) [reply pkg/drawio/iconTreeNode.go:8]: done, also there is an issue on it

File: pkg/drawio/layout.go (hunk 6 above)
Comment (haim-kermany) [reply pkg/drawio/layout.go:139]: there are some methods in layout that has the receiver (ly *layoutS) but do not use it.
conceptually they belong to layoutS, static methods.
is there another way to say that they are part of layoutS?

File: pkg/drawio/treeNodeInterface.go (hunk 2 above)
Comment (zivnevo) [reply pkg/drawio/treeNodeInterface.go]: I don't see it fixed. Can you please recheck?

File: pkg/drawio/createMapFile.go (hunk 1 above)
Comment (zivnevo) [reply pkg/drawio/createMapFile.go]: I'm probably missing something. Why can't `canTypeHaveAMiniIcon` be a global map like `decoreStyles` or the other maps in `styles.go`?

File: pkg/drawio/layout.go (hunk 6 above)
Comment (haim-kermany) [reply pkg/drawio/layout.go]: done

File: pkg/drawio/createMapFile.go (hunk 1 above)
Comment (haim-kermany) [reply pkg/drawio/createMapFile.go]: // if there are no vsi icon in the canvas, all the ni are displayed as vsi icon, and without mini icons
i.e - if there is no vsi icon, than canTypeHaveAMiniIcon[ni] will be false

File: pkg/drawio/treeNodeInterface.go (hunk 2 above)
Comment (haim-kermany) [reply pkg/drawio/treeNodeInterface.go]: fixes

File: pkg/drawio/createMapFile.go (hunk 1 above)
Comment (zivnevo) [reply pkg/drawio/createMapFile.go]: Got it

**Review by zivnevo — APPROVED** [review]
